src/jsn.erl

```diff
@@ -357,6 +359,71 @@ path_elements(_Type, _Path, _Acc) -> erlang:error(badarg).
 
 
 %%==============================================================================
+%% select API
+%%==============================================================================
+
+%%------------------------------------------------------------------------------
+%% @doc Transforms a list of Elements according to a given OutputSpecification
+%%------------------------------------------------------------------------------
+-spec select(selection(), json_array()) -> json_array().
+select(OutputSpec, Elements) when is_list(Elements) ->
+    select(OutputSpec, [], Elements);
+select(_OutpuSpec, _Elements) ->
+    erlang:error(badarg).
+
+%%------------------------------------------------------------------------------
+%% @doc Same as select/2 but it first filters the elements according to a given
+%% list of Conditions
+%%------------------------------------------------------------------------------
+-spec select(selection(), conditions(), json_array()) -> json_array().
+select(OutputSpec, Conditions, Elements) when is_list(Elements) ->
+    %% ensures Conditions are a list
+    ConditionList =
+        case Conditions of
+            _ when is_list(Conditions) -> Conditions;
+            _ -> [Conditions]
+        end,
+    %% This function applies a single condition to a single element
+    ApplyConditionFun = 
+        fun({Path, FilterFun}, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [?MODULE:get(Path, E)]);
+           ({Path, Value}, E) ->
+                ?MODULE:get(Path, E) == Value;
+           (FilterFun, E) when is_function(FilterFun) ->
+                erlang:apply(FilterFun, [E]);
+           (_, _) ->
+                erlang:error(badarg)
```

Comment:
I think it might be more client friendly to do something like this here:

```erlang
throw({error, {invalid_condition, Condition}}
```

I think that a `badarg` this deep into the top-level (client-facing) function is likely to result in a stack trace that is *very* hard to understand.

Maybe if you refactor this to not use anonymous functions, though, that won't be an issue.

Reply:
I agree that the error `badarg` here would be pretty opaque, but I'm not necessarily sure it needs to be thrown. What's the behavior in other locations when invalid parameters that need validation are passed into functions?  The behavior here should be the same.